### ISSUE TITLE
Plumb error messages from sdk resolution through hostfxr_resolve_sdk2

### ISSUE THREAD

**Issue by nguerrera** [dotnet/core-setup#6970]
sdk resolution emits errors to stderr, which is sufficient for command line scenarios, but stderr is not captured by VS when resolving SDKs.

(Aside: For a long time, VS would not show the errors returned by an SDK resolver at all, but that was fixed as of VS 16.0)

As of now, the that is shown in VS when sdk resolution fails is overly generic:

https://github.com/dotnet/cli/blob/6b0885139c1f37093ba4d3797f46fe90bf093883/src/Microsoft.DotNet.MSBuildSdkResolver/Strings.resx#L133

> Unable to locate the .NET Core SDK. Check that it is installed and that the version specified in global.json (if any) matches the installed version.

#### Proposed solution:

1. Add keys for error and warning to https://github.com/dotnet/core-setup/blob/1a10af88e3904b1d9790be38450c23eea683d6ae/src/corehost/cli/fxr/hostfxr.cpp#L141-L145

2. Change sdk resolver to retain the errors and warnings that are logged  only to stderr, and hand them back to callers of hostfxr_resolve_sdk2

3. Change msbuild sdk resolver to return these back to msbuild / VS.

https://github.com/dotnet/cli/blob/6b0885139c1f37093ba4d3797f46fe90bf093883/src/Microsoft.DotNet.MSBuildSdkResolver/NETCoreSdkResolver.cs#L25

cc @peterhuene @vitek-karas @KathleenDollard

**Comment by vitek-karas** [comment]
All the hosting layers already have a solution for this. In `hostfxr` it's here:
https://github.com/dotnet/core-setup/blob/1a10af88e3904b1d9790be38450c23eea683d6ae/src/corehost/cli/fxr/hostfxr.cpp#L391

The `hostfxr_set_error_writer` sets a thread-local variable which if set means all errors are written to the supplied callback and are not output to stderr. It's the responsibility of the custom error writer to report those errors in some way.

The caller should always remember the previous writer (returned by the set function) and set it back once the operation is done.
`hostfxr` would propagate the writer to `hostpolicy` if it makes calls into it as part of the operation, so all hosting errors are reported through the writer.

**Comment by vitek-karas** [comment]
We don't have any handling like this for warnings, but we could add something if necessary.

**Comment by nguerrera** [comment]
Oh, cool! I didn't know that was added. (I don't think it was there when hostfxr_resolve_sdk2 was added.)

I will close this and replace it with two different ones:

1. Allow receiving warnings (at which point I wonder if it should be generalized to all levels).

2. Use hostfxr_set_error_writer in msbuild sdk resolver. This will go in dotnet/cli.

I would say warnings are much lower priority.  So we can do (2) first and see if we really need them.

**Comment by nguerrera** [comment]
Filed:

1. https://github.com/dotnet/core-setup/issues/6972
2. https://github.com/dotnet/cli/issues/11651
